Issue #8222 - Addressing NPE with JvmArgs and non-existent modules.

joakime opened the pull request:
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #8222 - Addressing NPE with JvmArgs and non-existent modules.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #8222 - Addressing NPE with JvmArgs and non-existent modules.

+ Better variable naming.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented on jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java:
```diff
@@ -646,11 +648,27 @@ public Classpath getClasspath()
+     * Filtering out any that are invalid (such as modules that were not, or cannot be enabled).
+     * @return the stream of {@link Module} instances
+     */
+    public Stream<Module> getValidEnabledModulesStream()
```
This is a convenience method that makes usage predictable.
This addressed the NPE directly.

joakime replied:
No longer present

joakime commented on jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java:
```diff
@@ -646,11 +648,27 @@ public Classpath getClasspath()
+    /**
+     * The list of selected Modules to enable based on configuration.
+     * @return the list of selected modules (by name) that the configuration is requesting.
+     */
```
Added javadoc to explain what this list actually is.

joakime commented on jetty-start/src/main/java/org/eclipse/jetty/start/Main.java:
```diff
@@ -325,12 +326,22 @@ public StartArgs processCommandLine(String[] cmdLine) throws Exception
+        List<String> selectedModules = args.getEnabledModules();
+        List<String> sortedSelectedModules = modules.getSortedNames(selectedModules);
+        List<String> unknownModules = new ArrayList<>(selectedModules);
+        unknownModules.removeAll(sortedSelectedModules);
```
The use of `modules.getSortedNames(List<String>)` actually left out unknown modules from the returned list.
This filtering to show unknown modules is how we now error out when an unknown module is selected to be enabled.

joakime commented on jetty-start/src/main/java/org/eclipse/jetty/start/Main.java:
```diff
@@ -470,8 +481,7 @@ else if (args.isCreateFiles() || !args.getStartModules().isEmpty())
 
-            List<String> execModules = args.getEnabledModules().stream()
-                .map(name -> args.getAllModules().get(name))
+            List<String> execModules = args.getValidEnabledModulesStream()
```
Direct prevention of the NPE.
This is preventative, ideally we shouldn't be able to reach here with bad module names, as it should error out with a UsageException for modules that don't exist earlier.

gregw replied:
This feels a bit artificial.  I'd prefer that we just modified `getEnabledModules` to never return null.

joakime replied:
That's just it.

`getEnabledModules()` returns what the user requested to be enabled. aka: what they selected.
As the javadoc now says, this is the list of all `--module=<name>` entries in their configuration.
This is what they have asked for, their configuration, not what has actually been enabled. (which is often this selected list and the transitive modules)

IMO, we should not change that behavior.
It's also useful for the error message later, to show the whole list of unknown module names.

joakime replied:
This should have been called something else, "getSelectedModuleNames" or "getRequestedModuleNames".
It's not the list of enabled modules, never has been.
We don't even update this list for full list of modules that includes the transitive modules.
That's what the `Modules` object is for.

joakime replied:
The new method `StartArgs.getValidEnabledModulesStream() `is no longer present.

gregw commented on jetty-start/src/main/java/org/eclipse/jetty/start/Main.java:
```diff
@@ -325,12 +326,22 @@ public StartArgs processCommandLine(String[] cmdLine) throws Exception
 
         // 4) Active Module Resolution
-        for (String enabledModule : modules.getSortedNames(args.getEnabledModules()))
+        List<String> selectedModules = args.getEnabledModules();
```
To reduce confusion, can we keep to one verb. Use either selected or enabled, but not both.
```suggestion
        List<String> enabledModules = args.getEnabledModules();
```

joakime replied:
See prior comment about what this method `StartArgs.getEnabledModules()` actually does. (also see javadoc)

gregw replied:
When this variable is used, it is always the list of enabled modules.   The `sortedSelectedModules` and the `unknownModules` collections are created from what should be called `enabledModules`.

gregw reviewed (changes requested):
mostly good, just a few niggles

joakime committed:
Issue #8222 - Addressing NPE with JvmArgs and non-existent modules.

+ Simpler use case, without need for new StartArgs method

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-start/src/main/java/org/eclipse/jetty/start/Main.java:
```diff
@@ -470,8 +481,7 @@ else if (args.isCreateFiles() || !args.getStartModules().isEmpty())
 
-            List<String> execModules = args.getEnabledModules().stream()
-                .map(name -> args.getAllModules().get(name))
+            List<String> execModules = args.getAllModules().getEnabled().stream()
```
How is `args.getAllModules().getEnabled().stream()` different from `args.getEnabledModules().stream()`?

If it is different, then we have a naming problem.  If it is not different then just use `getEnabledModules()`

joakime replied:
I don't see this as a naming difference, just a "source of truth" difference.   
The `Modules` vs `StartArgs` are difference classes, with different purposes.

gregw reviewed (changes requested):
@joakime I'm not suggesting that we change behaviour.   I'm saying that we need to cleanup the implementation and/or the naming.

What is the difference between a "selected" module and an "enabled" module?
What is the difference between a module returned by `startArgs.getEnabledModule` and a module in `getAllModules` that returns true from `isEnabled`?  If there is no difference, then we should avoid duplicating state in the implemenation.  If there is a difference, then naming needs to reflect that difference.

If we can determine the set of enabled modules by simply filtering all modules on `isEnabled`, then do we actually need to maintain a collection of enabled modules in StartArgs?   If so, shouldn't it be called `enabledModules` rather than just `modules` etc.

I know this is a bit of scope increase for fixing just a NPE, but whilst we are looking at this, let's do it properly.

joakime commented:
`StartArgs` holds the configuration, in other words, what's in the `start.ini` and `start.d/*.ini` and command line.

It could, for example only have in it's `.getEnabledModules()` a List of type String with 2 entries `[ "http", "deploy" ]`.

The `Modules` class holds what has been enabled.
It's `Modules.getEnabled()` method (pre-existing, already there and doing the right thing) is a list of `Module` classes, which will return the complete list of enabled modules, based on the limited set provider in `StartArgs`, including all transitively enabled modules.
For the above `StartArgs` (`[ "http", "deploy" ]`), the `Modules.getEnabled()` will return the list `[bytebufferpool, resources, logging/slf4j, logging-jetty, threadpool, server, security, servlet, webapp, deploy, http]`.

The `StartArgs.getEnabledModules()` is the wrong API to use to build out the JvmArgs in the `Main` codepath for forked JVMs, as it will exclude any module specified JvmArgs as well.

The old, bad code, used `StartArgs.getEnabledModules()` in a stream, then mapped to a `Module`, then grabbed the JvmArgs, then filtered based on if the Module has JvmArgs, then mapped back to the name, to then get a list of module names used in the warning message for why start.jar is forking the JVM.

The new code, uses existing API properly here. (in Main for start with a forked JVM)
AND it also fails start up (with a UsageException) if a module name that doesn't exist is referenced with a complete list of referenced module names. (in Main for processCommandLine)
This last part used to be there, in the `Modules.enable(String, String)`, but the introduction of `modules.getSortedNames()` broke that location from triggering the UsageException.

gregw commented:
@joakime don't explain here. Explain in the code!   If  `StartArgs.getEnabledModules()` is the wrong API, the remove it or rename name it or at least document how it is different from `getAllModules().getEnabled()`.

Perhaps it is the case that the modules from start.ini and start.d/*.ini and command line should be called "selected" or "configured" rather than "enabled".    A module become enabled because it is selected/configured or is a dependency of a selected/configured module.

As I keep saying, I think the new code and handling of unknown modules is good.... but let's finish cleaning up the old code as well rather than leave two different concepts of what "enabled" means.

joakime committed:
Issue #8222 - Addressing NPE with JvmArgs and non-existent modules.

+ Correct naming in StartArgs.getEnabledModules
+ Introduce StartArgs.getSelectedModules
+ Change javadoc and usage everywhere else

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Stabilize flaky test by using UTC properly

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented:
@gregw done, code changed, javadoc added.

joakime commented on jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java:
```diff
@@ -71,7 +71,7 @@ public void testIsExpiredTimestampNextWeek()
     public void testIsExpiredTimestampNow()
     {
-        LocalDateTime now = LocalDateTime.now();
+        LocalDateTime now = LocalDateTime.now(ZoneId.of("UTC"));
```
This was a flaky test.
I just fixed it here as part of this PR.
It's really unrelated to the purpose of this PR, just happens to be in the same `jetty-start` code tree.

gregw reviewed (approved):
Much better!  thanks
